cypress-io/cypress
# feat(app): show "connect project" after login if project is not connected

- Closes #23022

### User facing changelog


If the user logs in from the App, but the current project does not have a project ID, the "Continue" button will we replaced with a "Connect project" button in the login confirmation modal. Clicking the "Connect project" modal will move the user to the Connect Project flow. Closing the modal with the X or by clicking away will _not_ have this effect, the user will be back in the app as before.

### Additional details


This change helps users of the Dashboard successfully connect their project after they have logged into the Dashboard into the app, since logging in does not provide a benefit unless the project is actually connected.

### Notes on implementation

This is is mostly driven by props and events. The Login Modal is a shared component between the App and the Launchpad, but the "Connect Project" flow is only currently possible in the App, so we are using a prop to tell the Login Modal whether it is allowed to show the "Connect project" button at that end, and letting the HeaderBarContent component determine whether it is in the app or the launchpad - which is something it already infers in deciding what content to show in the header.

This could be cleaned up more (little bit of "prop drilling"), but the next step for this feature is to move the Connect Project modal and related code into `frontend-shared` the props will go away, making it possible to connect a project from anywhere if a project is open. As it stands, the props/events make this easy to test.

### Alternatives considered

Simply placing the existing CloudConnectButton into the LoginModal and moving everything to shared seemed like it could be a quick win, but there ends up being a circular dependency since the CloudConnect flow also contains a LoginModal, so this led to errors.

The next ticket for this behavior will be to sort that out into more of a system, since really, if the LoginModal moves you to connect if necessary, it's possible to combine these components. Just a bigger change than needed for the immediate goal, so I'll make a new ticket for it. 

### Steps to test


- Open any Cypress project
- Remove the `projectId` from `cypress.config.ts` if there is one
- Log out if you are logged in
- Log in through the **header in the app**
  - See the "Connect project" button
  - Click the button and see the "create project" modal open
- Log out, log back in through the tooltips on the **specs list page**
  - See the "Connect project" button
  - Click the button and see the "create project" modal open
- Log out, log back in through **the Runs page** "Log in to the Cypress Dashboard" button
  - See the "Connect project" button
  - Click the button and see the "create project" modal open
- Re-add the `projectId` or add a new one in your config
- Log in through the same places and see that the button says Continue
- Log in through the Launchpad and see that the button never says "Connect project" even if you don't have a project id set. 


### How has the user experience changed?


New state after login if a project is not connected:

<img width="520" alt="Screen Shot 2022-08-11 at 10 26 34 PM" src="https://user-images.githubusercontent.com/8340719/184368532-6cd5a56e-b0be-4344-938e-b1b3c7758cdd.png">

### PR Tasks


- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
